libcaf_core/caf/flow/observable.test.cpp

```diff
@@ -249,6 +249,83 @@ TEST("on_error_complete() suppresses errors") {
   }
 }
 
+TEST("on_error_return_item() returns item on error") {
+  SECTION("on_error_return_item() does nothing if no error occurs") {
+    SECTION("blueprint") {
+      check_eq(collect(range(1, 1).on_error_return_item(42)), vector{1});
+      check_eq(collect(range(1, 2).on_error_return_item(42)), vector{1, 2});
+      check_eq(collect(range(1, 3).on_error_return_item(42)), vector{1, 2, 3});
+      check_eq(collect(range(1, 0).on_error_return_item(42)), nil);
+    }
+    SECTION("observable") {
+      check_eq(collect(mat(range(1, 1)).on_error_return_item(42)), vector{1});
+      check_eq(collect(mat(range(1, 2)).on_error_return_item(42)),
+               vector{1, 2});
+      check_eq(collect(mat(range(1, 3)).on_error_return_item(42)),
+               vector{1, 2, 3});
+      check_eq(collect(mat(range(1, 0)).on_error_return_item(42)), nil);
+    }
+  }
+  SECTION("on_error_return_item() returns item when an error occurs") {
+    SECTION("blueprint") {
+      check_eq(collect(range(1, 0)
+                         .concat(obs_error(sec::runtime_error))
```

Comment:
I had to use the 0 size range to ensure that the value is returned properly. The `fail` operator does not return items and an empty vector is returned when `on_complete` is called. If this is not the expected behavior, I think that we need to update the `fail` operator accordingly. https://github.com/actor-framework/actor-framework/blob/93d522fc51110c77320c27f452ab7a3417a80aed/libcaf_core/caf/flow/op/fail.hpp#L30-L33

Reply:
> I had to use the 0 size range to ensure that the value is returned properly.

Now I'm confused. What value? `range(1, 0).concat(obs_error())` should be identical to just `obs_error()`, no? `range(1, 0)` should call `on_complete()` immediately (since we asked it to produce 0 values), then, `concat(...)` will switch to the next observable, which calls `on_error()` immediately. End result should be an observable that simply calls `on_error()`... which is exactly what just using `obs_error()` alone achieves.

Reply:
`range(1, 0).concat(obs_error())` is not completely identical to `obs_error()` in terms of output. Though `range(1, 0)` calls on_complete immediately, `from_generator.hpp` calls `out_.on_next()` in it's `do_run` method which ultimately calls `from_steps_sub::on_next()`. If this function is not called, `from_steps_sub::do_run()` seems to return an empty array even though the `buf_` is not empty. I think that this maybe a bug with how `fail` operator is created. Shall I attempt to fix this in the PR as well?